### PR TITLE
feat: add partial tablet v2 protocol support

### PR DESCRIPTION
**Pull request by jp7677**
As a follow up to https://github.com/labwc/labwc/issues/1672 and the ideas mentioned there, I ~played/hacked~ put this together. Running `gtk4-demo` and there the `Paint` demo, I can draw now in Stylus only mode and pen pressure is applied. So far so good.

~Looking at the general UI, just blindly switching from mouse emulation to tabled mode disables a lot of things that are imho not so easy to get correctly implemented, e.g. focus switch between applications, resize applications, drag & drop and surely a few others. That said, having a configuration switch like 'tablet-mode' (tablet-native|mouse-emulation) and/or an action 'toogle-tablet-mode' might be a way to mitigate this and allow a user to consciously switch between modes with all the consequences and restrictions applied.~ Actually this is now in a very usable state, see below. A switch for mouse emulation is also part of this PR.

~There are more tablet axis that aren't covered yet in this PR.~ I also haven’t done much testing with side businesses like disconnecting/reconnecting the tablet, multiple outputs, output mapping.

So, open for feedback (esp. with a real drawing app like Krita) and thoughts.

Status / TODO:

- [x] Tablet tool wiring/signalling 
- [x] Axis/proximity handling
- [x] Hovering between tablet-capable / non-tablet-capable surfaces / desktop
- [x]  Focus switching
- [x] Idle inhibit
- [x] Client cursors (set_cursor), tricky part...
- [x] Move/resize CSD applications
- [x] Out of surface scrolling
- [x] Button mapping (for non-tip)
- [x] Force-cursor-emulation configuration option
- [x] Add a few safety checks like tool type
- [ ] Drag n drop (if this turns out to be doable)
- [x] Document force-mouse-emulation and button mapping implications
- [x] Testing with device reconnects 
- [x] Testing with output mapping
- [ ] General testing
- [ ] Review, polish and reorganize commits
 
Currently known issues:
- Notify cursor from pointer and tablet can still interfere, noted on out-of-surface scrolling in GTK4 applications. The resize cursor shines through when scrolling out of surface.
- ~The scrollbar in VS-Code, when running as native Wayland application, only when being adjacent or above another window, jumps to the top when scrolling out of surface. This does not happen with VS-Code in XWayland or when the out of scrolling happens on the desktop.~ A similar issue can be seen with the volume slider in Shortwave/GTK4 (Slider on a popup, thus also scrolling out of surface above another surface).

Ideas:
- ~`wlr_tablet_v2_tablet_tool` has a `focused_surface` field which is updated on proximity-in/out.  May be we can use that one instead of our own `drawing_tablet_tool.focus_surface`.~ (Done)
-  Notify cursor interference could be solved by an early return in `request_cursor_notify` based on `drawing_tablet_tool.focus_surface`. Unfortunately that means we need to add and maintain the tool, the `focused_surface` or just a new `boolean tablet_tool_in_proximity` in `seat`.

### CODE REVIEW

File: src/input/tablet.c [hunk 1]
```diff
@@ -63,27 +66,97 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
...
+	struct wlr_surface *surface = tablet_get_coords(tablet, x, y, &nx, &ny);
+
+	if (surface && tablet->tablet_v2_tool) {
+		wlr_tablet_v2_tablet_tool_notify_proximity_in(
+			tablet->tablet_v2_tool, tablet->tablet_v2, surface);
+
+		if (ev->updated_axes & (WLR_TABLET_TOOL_AXIS_X | WLR_TABLET_TOOL_AXIS_Y)) {
+			// Visualize where we are and what we are
+			wlr_cursor_warp_absolute(tablet->seat->cursor, tablet->base, x, y);
+			cursor_set(tablet->seat, LAB_CURSOR_PENCIL);
```
Comment (jp7677) [inline src/input/tablet.c:121]: Most likely not the best idea ..

File: src/input/tablet.c [hunk 2]
```diff
@@ -63,27 +66,163 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
...
+
+	double lx, ly;
+	struct wlr_surface *surface = tablet_get_coords(tablet, x, y, &lx, &ly);
+
+	if (surface) {
+		wlr_log(WLR_DEBUG, "  got proximity event on surface");
+		// TODO: This seems just plain wrong and causes crashes
+		if (!tablet->tablet_v2_tool || tablet->tablet_v2_tool->wlr_tool != ev->tool) {
+			// Unfortunately I found `wlr_tool` only in the events,
+			// so use this for creating a `wlr_tablet_v2_tablet_tool`
```
Comment (Consolatis) [inline src/input/tablet.c:205]: `struct wlr_tablet_pad` appears to have a signal for attaching a tool (`attach_tablet`).

File: src/input/tablet.c (hunk 1 above)
Comment (Consolatis) [reply src/input/tablet.c:138]: I don't think we actually need to set the cursor manually here at all.
`struct wlr_tablet_v2_tablet_tool` has a `set_cursor()` signal though that supplies a wl_surface, so I assume we can handle it similar to `request_cursor_notify()` in `src/input/cursor.c`.

File: src/input/tablet.c [hunk 3]
```diff
@@ -63,27 +66,183 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
...
+				ev->distance);
+		}
+		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_PRESSURE) {
+			wlr_tablet_v2_tablet_tool_notify_pressure(tablet->tablet_v2_tool,
+				ev->pressure);
+		}
+		if (ev->updated_axes
+				& (WLR_TABLET_TOOL_AXIS_TILT_X | WLR_TABLET_TOOL_AXIS_TILT_Y)) {
+			wlr_tablet_v2_tablet_tool_notify_tilt(tablet->tablet_v2_tool,
+				tablet->tilt_x, tablet->tilt_y);
```
Comment (jp7677) [inline src/input/tablet.c:178]: Do we need to transform (area, rotation) tilt?

File: src/input/tablet.c [hunk 4]
```diff
@@ -63,27 +66,183 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
...
+		wlr_log(WLR_DEBUG, "  got proximity event on surface");
+		// TODO: This seems just plain wrong and causes crashes
+		if (!tablet->tablet_v2_tool || tablet->tablet_v2_tool->wlr_tool != ev->tool) {
+			// Unfortunately I found `wlr_tool` only in the events,
+			// so use this for creating a `wlr_tablet_v2_tablet_tool`
+			wlr_log(WLR_DEBUG, "  creating tablet_v2_tool");
+			if (tablet->tablet_v2_tool) {
+				wlr_log(WLR_DEBUG, "  tablet_v2_tool was already created");
+			}
+			tablet->tablet_v2_tool =
```
Comment (jp7677) [inline src/input/tablet.c:226]: Free somewhere?

File: src/input/tablet.c [hunk 5]
```diff
@@ -133,13 +311,23 @@ tablet_init(struct seat *seat, struct wlr_input_device *wlr_device)
 	wlr_log(WLR_DEBUG, "setting up tablet");
 	struct drawing_tablet *tablet = znew(*tablet);
 	tablet->seat = seat;
+	tablet->base = wlr_device;
 	tablet->tablet = wlr_tablet_from_input_device(wlr_device);
 	tablet->tablet->data = tablet;
+	if (seat->server->tablet_manager) {
+		tablet->tablet_v2 = wlr_tablet_create(
```
Comment (jp7677) [inline src/input/tablet.c:318]: Free somewhere?

File: src/input/tablet_tool.c [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		wl_container_of(listener, tool, handlers.set_cursor);
+	struct wlr_tablet_v2_event_cursor *ev = data;
+
+	struct seat *seat = tool->tablet->seat;
+
+	if (seat->server->input_mode != LAB_INPUT_STATE_PASSTHROUGH) {
+		return;
+	}
+
+	wl_list_remove(&tool->handlers.surface_destroy.link);
```
Comment (jp7677) [inline src/input/tablet_tool.c:52]: Do I need to guard this against concurrent calls from multiple client?

File: src/input/tablet_tool.c [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		ev->hotspot_x, ev->hotspot_y);
+}
+
+static void
+handle_destroy(struct wl_listener *listener, void *data)
+{
+	struct drawing_tablet_tool *tool =
+		wl_container_of(listener, tool, handlers.destroy);
+
+	wl_list_remove(&tool->link);
```
Comment (jp7677) [inline src/input/tablet_tool.c:70]: Also clean up listeners?

File: src/input/tablet_tool.c (hunk 6 above)
Comment (johanmalm) [reply src/input/tablet_tool.c:52]: I don't know, but it feels like only the 'focused' client should be allowed to request `set_cursor` (just guessing though).

File: src/input/tablet_tool.c (hunk 7 above)
Comment (johanmalm) [reply src/input/tablet_tool.c:70]: Yes, most likely. `set_cursor` feels like a certainty. Not sure about `proximity`, etc - possible they get destroyed by the manager.
Might be worth doing some printf()s to make sure it balances. Always a tricky area to get right.

File: src/input/tablet_tool.c [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+#include "config/rcxml.h"
+#include "input/cursor.h"
+#include "input/tablet_tool.h"
+#include "labwc.h"
+
+void
+tablet_tool_cursor_reset(struct drawing_tablet_tool *tool)
+{
+	cursor_set(tool->seat, LAB_CURSOR_DEFAULT);
+	cursor_update_image(tool->seat);
```
Comment (jp7677) [inline src/input/tablet_tool.c:18]: note to self: probably no longer needed?

File: src/input/tablet.c (hunk 2 above)
Comment (Consolatis) [reply src/input/tablet.c]: I looked a bit into wlroots and the issue is that wlroots destroys the tool on proximity_out when it can't be uniquely identified. So using the proximity_in signal to create the tool wrapper seems like the correct solution.

**Comment by Consolatis** [comment]
Managed to somewhat fix my test setup (the battery in the pen was leaking and corroded some contacts) and this seems to work great. Only issue I saw while testing is that I was unable to resize the `gtk4-demo --run=paint` window with the pen. I am not quite sure why that is though, it might be a GTK bug in their tablet input handling. Usually CSD clients should request entering interactive resize mode from the compositor when a client clicks on the invisible area around the window but that doesn't appear to happen when the input comes from a tablet. Same thing for moving the (CSD) window when dragging the titlebar.

I am also not sure about the cursor surface destroy listener, we do not use it in `src/cursor.c` either and I'd assume (not actually verified) that wlroots handles that internally. Another thing is that we IMHO should not hardcode the focus behavior here but rather call the same user defined action as when pressing the left mouse button.

**Comment by jp7677** [comment]
> Managed to somewhat fix my test setup (the battery in the pen was leaking and corroded some contacts) and this seems to work great. 

Thanks a lot for starting to look into this and cool to hear that you got your tablet working again.

> Only issue I saw while testing is that I was unable to resize / move the `gtk4-demo --run=paint` window with the pen. I am not quite sure why that is though, it might be a GTK bug in their tablet input handling. Usually CSD clients should request entering interactive resize mode from the compositor when a client clicks on the invisible area around the window but that doesn't appear to happen when the input comes from a tablet. Same thing for moving the (CSD) window when dragging the titlebar.

Yeah, I haven't figured yet what is need for resizing or moving CSD applications, I can't resize/move GTK3 nor GTK4 applications. I'm somehow convinced that _I'm_ missing something ;). I've already tried adding `wlr_tablet_tool_v2_start_implicit_grab()` to tip-down, but that didn't do the trick.

> I am also not sure about the cursor surface destroy listener, we do not use it in `src/input/cursor.c` either and I'd assume (not actually verified) that wlroots handles that internally. 

`wlroots` adds its own listener which effectively hides the cursor on surface destroy (https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/wlr_cursor.c#L665 / https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/wlr_cursor.c#L633). I've added this to prevent the behavior described in https://github.com/labwc/labwc/issues/1672#issuecomment-2045775374 . But yeah, I should try again without because I've changed a few things in the meantime.

> Another thing is that we IMHO should not hardcode the focus behavior here but rather call the same user defined action as when pressing the left mouse button, similar to what is done in [touch.c](https://github.com/labwc/labwc/blob/master/src/input/touch.c#L115).

Ah right, I can do that instead of `desktop_focus_view`.

**Comment by Consolatis** [comment]
> > I am also not sure about the cursor surface destroy listener, we do not use it in `src/input/cursor.c` either and I'd assume (not actually verified) that wlroots handles that internally.
> 
> `wlroots` adds its own listener which effectively hides the cursor on surface destroy (https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/wlr_cursor.c#L665 / https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/wlr_cursor.c#L633). I've added this to prevent the behavior described in [#1672 (comment)](https://github.com/labwc/labwc/issues/1672#issuecomment-2045775374) . But yeah, I should try again without because I've changed a few things in the meantime.

Assuming you were referring to this part of the comment:

> I've noticed that when the last window of Firefox or Thunderbird gets closed and the cursor falls to the desktop then it will disappear until it hovers a new window. I couldn't reproduce it with any other GTK/Qt/Xwayland app.

As long as the cursor is set to `LAB_CURSOR_CLIENT` the `cursor_update_focus()` call on view destruction should handle that.

**Comment by jp7677** [comment]
> As long as `seat->server_cursor` is set to `LAB_CURSOR_CLIENT` the `cursor_update_focus()` call on view destruction should handle that.

yes, it was needed anymore. Nice!

File: include/labwc.h [hunk 9]
```diff
@@ -114,6 +115,13 @@ struct seat {
 
 	struct wlr_pointer_constraint_v1 *current_constraint;
 
+	/*
+	 * Tracks the top-most surface underneath a tablet tool when the
+	 * tool is in proximity.
+	 * Updating happens on tablet proximity and tablet axis events.
+	 */
+	struct wlr_surface *tablet_tool_surface;
```
Comment (jp7677) [inline include/labwc.h:123]: note to self: might also track the tool here instead, may be that allows a cleaner plumbing of the tablet functions into the general cursor logic.

**Comment by jp7677** [comment]
@Consolatis I've managed to integrate the tablet handling into the existing cursor handling by exposing a few general methods that return if a notify (pointer or tablet) should follow. From my testing and after a few iterations this does the job quite good, even exotic things like continue scrolling with moving the stylus out of surface works. That said, the changes I made to the cursor functions are a bit hard to digest and certainly introduced complexity and ugliness. On the plus  side, it might be feasible now to extract the pointer specific signal handling from `cursor.c` into a another file for better separation. 
Please let me know what you think of that direction and also ofc if you have an inspiration on how to merge the pointer and tablet stuff more elegantly.
PS: The commits are currently somewhat messy, I'll clean this up once finished.

File: src/input/cursor.c [hunk 10]
```diff
@@ -140,6 +140,14 @@ request_cursor_notify(struct wl_listener *listener, void *data)
 		return;
 	}
 
+	if (!wl_list_empty(&seat->server->tablet_manager->clients)) {
```
Comment (jp7677) [inline src/input/cursor.c:143]: note to self: TODO: this prevents the drag cursor going back to a normal cursor when finishing moving a CSD application.

Comment (jp7677) [reply src/input/cursor.c]: Remove that part. While it fixed the hovering into a GTK4 app that wants a different cursor when entering its surface (thus when somewhat still being over de resize area), it also prevented setting a cursor in XWayland apps without tablet support (e.g. VS Code) next to the remaining cursor after moving.

File: src/interactive.c [hunk 11]
```diff
@@ -251,4 +251,5 @@ interactive_cancel(struct view *view)
 
 	/* Update focus/cursor image */
 	cursor_update_focus(view->server);
+	cursor_set(&view->server->seat, LAB_CURSOR_DEFAULT);
```
Comment (jp7677) [inline src/interactive.c:254]: note to self: probably controversial.

File: src/input/cursor.c (hunk 10 above)
Comment (jp7677) [reply src/input/cursor.c]: Re-added, somehow still makes sense to prevent those conflicts.Mitigated by resetting the cursor in interactive_cancel. Let's play some more.

**Comment by Consolatis** [comment]
> Edit: please don’t do a full code review yet, I need to e.g. polish and add code comments first, for now it’s about the direction of combining the cursor and tablet magic.

I think the general idea of having the cursor functions return a bool which is then used by the caller to either notify the client or not is sensible. It should likely be split into its own PR when everything works as intended for the tablet use-case (e.g. there aren't any missing pieces in the cursor part to make the tablet part make use if it). That makes the review a bit simpler.

**Comment by jp7677** [comment]
> I think the general idea of having the cursor functions return a bool which is then used by the caller to either notify the client or not is sensible. It should likely be split into its own PR when everything works as intended for the tablet use-case (e.g. there aren't any missing pieces in the cursor part to make the tablet part make use if it). That makes this PR (and its review) a bit simpler. 

Yes, good idea. I've already isolated that to a single commit https://github.com/labwc/labwc/pull/1678/commits/ccd715660758d4abfe9d44bfb9f481399c88d462 , should be trivial to make that a separate PR once everything is in place and confirmed to work.

> Overall I think we should not ship this feature in `0.7.2` (targeted feature deadline on April 25th).

Fully agree, I'm sure I'll still need a few weeks to get everything sorted ;)

**Comment by jp7677** [comment]
I'm running this PR now for daily use for a few days with no mouse attached and it works imho pretty good. A few minor quirks I have noticed:
- When moving or resizing the Steam window, the cursor (sometimes) in Steam itself doesn't change shape anymore, e.g. no hand cursor above links. The UI still reacts to the cursor and click works, just the shape is off. The cursor shape restores as soon as the stylus goes proximity out/in or hovering out/in again.
- When two application windows are side by side, the left one is CSD/GTK4 and the right one is focused, I cannot grab the scrollbar of the GTK4 application. Click on scrollbar does work and scrolls, but not grabbing the scrollbar. Once the GTK4 application is focused, grabbing the scrollbar works again. Grabbing works when the left application was e.g. Firefox or Tilix (both GTK3).

@rhyses-pieces experiences cursor flickering above XWayland applications, notably Krita, but this also happens in Sway and might be rooted outside of this PR. Furthermore a labwc restart is needed for restoring pen functionality after a reconnect (the display of the tablet does get connected again though). Both scenarios do work fine on my setup.

**Comment by jp7677** [comment]
> When two application windows are side by side, the left one is CSD/GTK4 and the right one is focused, I can grab the scrollbar of the GTK4 application but cannot move it (or just a tiny bit) . Sometimes it also doesn't work when the GTK4 app is focused. Click on scrollbar does work and scrolls, but not grabbing the scrollbar. Grabbing works when the left application is e.g. Firefox or Tilix (both GTK3).

I'm slightly lost with this issue. Scrolling works fine when no other application is adjacent to e.g. nautilus, but as soon as another application is (and that other application has focus), I cannot get hold of the scrollbar for as long as I hold tap down, only for a tiny moment. The behavior is fine when using a mouse or mouse emulation and also when using the stylus in Gnome. Any idea's?

**Comment by 01micko** [comment]
@jp7677 probably pay to ping @Consolatis @jlindgren90 @ahesford @tokyo4j :)

re https://github.com/labwc/labwc/pull/1678#issuecomment-2081349929

Er, done!

**Comment by johanmalm** [comment]
What you describe sounds familiar, but I've not looked at patch, so might be barking up the wrong tree.

When grabbing a scrollbar, you have to keep the associated surface focused even if the cursor moves out over other clients/surfaces.

Ref:
- https://github.com/labwc/labwc/blob/2603dbf396bddeebefeb8e6e1f4ef7cce3b4eb31/src/input/cursor.c#L327
- https://github.com/labwc/labwc/blob/2603dbf396bddeebefeb8e6e1f4ef7cce3b4eb31/src/input/cursor.c#L417-L424

**Comment by jp7677** [comment]
> What you describe sounds familiar, but I've not looked at patch, so might be barking up the wrong tree.
> 
> When grabbing a scrollbar, you have to keep the associated surface focused even if the cursor moves out over other clients/surfaces.

Thanks for the reply and sorry for the late response, I had been away from keyboard for a while. 

Actually I'm reusing all the common cursor logic you are referring to, so the tablet code hits all the code places you mentioned. I still don't know what exactly is causing this weird scrolling behavior I'm seeing when another application is adjacent to a GTK4 application. The scrolling behavior works perfectly when no other application is adjacent. Whatever I do code-wise, I cannot get the scrolling to work correctly when using tablet notifications. I tend to think that this is more a GTK4 quirk (their tablet protocol implementation i.c.w. focus behavior?), but dunno. I don't have this issue with e.g. GTK3 applications and, interestingly, enabling focus-follows-mouse also completely mitigates this issue.

**Comment by jp7677** [comment]
> When moving or resizing the Steam window, the cursor (sometimes) in Steam itself doesn't change shape anymore, e.g. no hand cursor above links. The UI still reacts to the cursor and click works, just the shape is off. The cursor shape restores as soon as the stylus goes proximity out/in or hovering out/in again.

This should be resolved now, I'm doing now an additional proximity-out/proximity-in after an interactive move/resize which seems to do the trick. With this I could also remove a few hacky lines in the last commit.

So only the GTK4 scrollbar quirk is left.

**Comment by jp7677** [comment]
> So only the GTK4 scrollbar quirk is left (based on my own testing).

I found a way to work around this issue, by doing additionally `wlr_seat_pointer_notify_motion(tool->seat->seat, time, 0, 0)` to `wlr_tablet_v2_tablet_tool_notify_motion()`, see last commit.  I haven't seen yet any negative side effects, but feels rather ugly. That said, from a functional perspective, all should be good now.

**Comment by johanmalm** [comment]
I'm only half way through the commit in terms of review.

Comments so far:

1. `69e3c36`: When calling `cursor_update_common()` (nx, ny) has been used rather than (sx, sy).
   See: https://github.com/labwc/labwc/blob/2f1fcb4468d1d6e3442a3cce2d13ea26f3a7b805/CONTRIBUTING.md?plain=1#L311
   
2. `6c22773` Same comment for `tablet_get_coords()`
  
3. `d020bf5` In `struct drawing_tablet` you have added `struct wlr_input_device *base;`. Normally wlroots (and us sometimes) use base to refer to a base struct within a derived struct in the OOP-inheritance sense. It allows getting the inherited struct from the base with constructs such as the one below. It doesn't look like you've used it in that way though, so `struct wlr_input_device *wlr_input_device;` is easier to read where it's dereferenced - and more consistent with code base.


```
struct input *input;
wl_list_for_each(input, &seat->inputs, link) {
  if (input->wlr_input_device->type == WLR_INPUT_DEVICE_KEYBOARD) {
    struct keyboard *keyboard = (struct keyboard *)input;
    ...
  }
}
```

- https://github.com/labwc/labwc/blob/2f1fcb4468d1d6e3442a3cce2d13ea26f3a7b805/src/seat.c#L30
- https://github.com/labwc/labwc/blob/2f1fcb4468d1d6e3442a3cce2d13ea26f3a7b805/include/labwc.h#L84

4. I don't get why we do `wlr_seat_pointer_notify_motion(tool->seat->seat, time, -1, -1);` (probably just me being a bit slow).

**Comment by jp7677** [comment]
> I'm only half way through the commit in terms of review.

Thanks a lot for already looking into the details!
 
> Comments so far:

I've addressed the renames in point 1, 2 and 3.

>  I don't get why we do `wlr_seat_pointer_notify_motion(tool->seat->seat, time, -1, -1);` (probably just me being a bit slow).

By re-using the existing cursor logic, this PR also sets pointer focus (`wlr_seat_pointer_notify_clear_focus` / `wlr_seat_pointer_notify_enter`) when a tablet pen enters a surface. This is a good thing. A side effect is though, that a client will be notified about a pointer event (pointer enter at coordinate) and tablet events (proximity-in, move to coordinate) at the same time. The following tablet actions, as long as the pen stays on the surface, emit only tablet events. That said, the client still thinks that there is a pointer at the coordinate where the tablet entered the surface. This can conflict with e.g. menu or scrollbars because the client might be conflicted about which coordinates (pointer or tablet) have priority when both coordinates are on the same menu or scrollbar.  I figured accidentally that -1/-1 are valid coordinates (well, in the sense that it works). Notifying pointer motion to -1/-1, the pointer is kind of out of the way and is never on the same control like the tablet pen, so no conflicts anymore for the client. 
Does this makes sense?

Note that Gnome/Mutter sends a pointer-leave notification on tablet proximity-in to the client to avoid this conflict. Going that way would probably involve much more refactoring in labwc and I'm not sure what will break since I have the feeling that a lot of internals rely on correct pointer focus.

**Comment by johanmalm** [comment]
Makes sense now. There is already a fairly length comment, but suggest adding some/all of the above for future devs :smile:

**Comment by jp7677** [comment]
> Makes sense now. There is already a fairly length comment, but suggest adding some/all of the above for future devs 😄

Thanks and done!

**Comment by jp7677** [comment]
I think this should be good now. Please let me know if i should open a separate PR for the first commit.

**Comment by johanmalm** [comment]
LGTM - but please don't merge yet.
The @Consolatis is also reviewing and they have a track-record of finding things!

**Comment by Consolatis** [comment]
This looks good to me as well. Didn't detect any issue while testing and overall feels very complete.

With the comment added as raised by malm I'd say lets go for the merge here.

I am not particularly happy about the cursor parking at -1 -1 but we can deal with that in another PR. Same story for drag and drop support. Thanks a lot for all the work you've put into this!

**Comment by jp7677** [comment]
> This looks good to me as well. Didn't detect any issue while testing and overall feels very complete. With the comment added as raised by malm I'd say lets go for the merge here.

Cool, I amended two commits with comments. I hope they make some sense.

> I am not particularly happy about the cursor parking at -1 -1 but we can deal with that in another PR. Same story for drag and drop support. Thanks a lot for all the work you've put into this!

It was quite a ride  ;)

File: src/input/tablet.c [hunk 12]
```diff
@@ -93,60 +177,175 @@ handle_axis(struct wl_listener *listener, void *data)
...
+		tablet->tilt_x = ev->tilt_x;
+	}
+	if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_TILT_Y) {
+		tablet->tilt_y = ev->tilt_y;
+	}
+
+	double x, y;
+	struct wlr_surface *surface = tablet_get_coords(tablet, &x, &y);
+
+	if (tool && ((surface
```
[the remark this answers is not in the document]
Comment (jp7677) [reply src/input/tablet.c]: Should be good now.

File: src/input/tablet.c [hunk 13]
```diff
@@ -75,77 +82,410 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
...
+{
+	idle_manager_notify_activity(tool->seat->seat);
+
+	bool enter_surface = false;
+	/* Postpone proximity-in on a new surface when the tip is down */
+	if (surface != tool->tool_v2->focused_surface && !tool->tool_v2->is_down) {
+		enter_surface = true;
+		wlr_tablet_v2_tablet_tool_notify_proximity_in(tool->tool_v2,
+			tablet->tablet_v2, surface);
+	}
```
Comment (tokyo4j) [inline src/input/tablet.c:129]: @jp7677 What's the reason behind `!tool->tool_v2->is_down`?

(I'm now reading `src/input/tablet.c` to refactor the entire logic of cursor handling and reuse `cursor_process_*` functions in `src/input/touch.c`).

Comment (jp7677) [reply src/input/tablet.c:129]: From the commit message of https://github.com/labwc/labwc/pull/1678/commits/594717d03b6043b6106854aa45bbd3d16059e819

`Also postpone proximity-in when moving to a new surface with the tip down.`

Its been a while, but i think it had to do with not interrupting drag&drop from not-tablet capable surfaces to tablet capable surfaces. 

Please be careful with the tablet stuff, it was far from trivial to get the tablet interaction right considering that the tablet protocol is completely separated from the normal pointer but somehow had to be bolted into it to allow normal desktop/window interactions and not just actual drawing.

Comment (tokyo4j) [reply src/input/tablet.c:129]: > Its been a while, but i think it had to do with not interrupting drag&drop from not-tablet capable surfaces to tablet capable surfaces.

But isn't it already handled in the check against `is_down_mouse_emulation`?

https://github.com/labwc/labwc/blob/7bf08af210968fb174149754ce341988365a131b/src/input/tablet.c#L372-L376

I don't know applications that support DnD and don't support tablet protocol, but at least out-of-surface dragging from a non-tablet-capable surface to a tablet-capable surface worked without the `!tool->tool_v2->is_down` check.

Please note that I never intend to blame you. I just want to make sure that my future changes don't break any tablet functionalities.

Comment (jp7677) [reply src/input/tablet.c:129]: Looking more closely, I think now that it was related to out-of-surface scrolling for a tablet-capable application when the out-surface area is covered by a second tablet-capable surface (e.g. out-of-surface scrolling in firefox above another GTK application). So to prevent a surface change for the tool (proximity out/in) as long as the out-of-surface scrolling is performed.

> Please note that I never intend to blame you. I just want to make sure that my future changes don't break any tablet functionalities.

No worries, all good. Thank you for asking in the first place and once we have the reason, it surely justifies a comment.

Comment (tokyo4j) [reply src/input/tablet.c:129]: I don't think the check for `tool->tool_v2->is_down` is needed for out-of-surface scrolling from a tablet-capable surface to another tablet-capable surface either, because we set implicit grab when the tip is down, which makes `wlr_tablet_v2_tablet_tool_notify_proximity_in()` no-op: https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/tablet_v2/wlr_tablet_v2_tool.c?ref_type=heads#L748

Comment (jp7677) [reply src/input/tablet.c:129]: I think I wasn't aware of this fact when writing that code. I think I just looked for a counterpart of `is_down_mouse_emulation` but from a tablet-capable surface. Sorry, I don't remember exactly anymore which scenario/edge case I had in mind back then and if I had tested without `is_down`.

Comment (tokyo4j) [reply src/input/tablet.c:129]: Thanks, that's totally fine. I was just wondering if I'm missing any edge-cases to test with.

I really appreciate your effort on this PR that made labwc more accessible to tablet users!
